### PR TITLE
Implement utility for dedicated tracing of well-known events

### PR DESCRIPTION
**Pull request by ggezer**

### CODE REVIEW

File: core/client/fs/src/main/java/alluxio/client/file/AlluxioEvent.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
+package alluxio.client.file;
```
Comment (gpang) [inline core/client/fs/src/main/java/alluxio/client/file/AlluxioEvent.java:1]: Why is this in "client"? Shouldn't this be in server?

Also, it probably also shouldn't be in "file" either, since these are really server level events, not file system events.

File: core/client/fs/src/main/java/alluxio/client/file/AlluxioEvent.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+ * Defines and Manages well-known Alluxio events.
+ *
+ * !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+ * !!! - Keep increasing the event-id for new events. !!!
+ * !!! - Don't reassign event ids.                    !!!
+ * !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+ */
+public enum AlluxioEvent {
+  // Alluxio master events.
+  Master_Process_Starting(1, EventType.INFO, "Master process starting."),
```
Comment (gpang) [inline core/client/fs/src/main/java/alluxio/client/file/AlluxioEvent.java:16]: I feel like these should either be camelcase, or all caps with underscores.

File: core/client/fs/src/main/java/alluxio/client/file/AlluxioEvent.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        LOG.info(eventString);
+        break;
+      case ERROR:
+        LOG.error(eventString);
+        break;
+      case WARNING:
+        LOG.warn(eventString);
+        break;
+      case DIAGNOSTIC:
+        LOG.debug(eventString);
```
Comment (gpang) [inline core/client/fs/src/main/java/alluxio/client/file/AlluxioEvent.java:64]: shouldn't the event log always be logging? Or is the idea that this log always enables debug?

File: core/client/fs/src/main/java/alluxio/client/file/AlluxioEvent.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  Master_Process_Stopping(2, EventType.INFO, "Master process stopping."),
+  // Alluxio worker events.
+  Worker_Process_Starting(3, EventType.INFO, "Worker process started."),
+  Worker_Process_Stopping(4, EventType.INFO, "Worker process stopping.");
+
+  // logger.
+  private static final Logger LOG = LoggerFactory.getLogger(AlluxioEvent.class);
+
+  /** The unique event id. */
+  private final int mId;
```
Comment (gpang) [inline core/client/fs/src/main/java/alluxio/client/file/AlluxioEvent.java:26]: What is the id for? Wouldn't the name of the enum be enough?

File: core/client/fs/src/main/java/alluxio/client/file/AlluxioEvent.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  /**
+   * Fires the event.
+   *
+   * @param args arguments to event message string
+   */
+  public void fire(Object... args) {
+    String msgString = String.format(mMsgFormat, args);
+    String eventString = String.format("Id: %d, Name: %s, Message:%s", mId, this.name(), msgString);
```
Comment (gpang) [inline core/client/fs/src/main/java/alluxio/client/file/AlluxioEvent.java:52]: Isn't the name of the enum enough of an id rather than the id int?

File: core/client/fs/src/main/java/alluxio/client/file/AlluxioEvent.java (hunk 2 above)
Comment (ggezer) [reply core/client/fs/src/main/java/alluxio/client/file/AlluxioEvent.java:16]: I tried all caps and it looked intrusive. 
I will try camel casing.

File: core/client/fs/src/main/java/alluxio/client/file/AlluxioEvent.java (hunk 4 above)
Comment (ggezer) [reply core/client/fs/src/main/java/alluxio/client/file/AlluxioEvent.java:26]: Numeric IDs are easier to communicate imo. So that's why I took the trouble to define it.

File: core/client/fs/src/main/java/alluxio/client/file/AlluxioEvent.java (hunk 3 above)
Comment (ggezer) [reply core/client/fs/src/main/java/alluxio/client/file/AlluxioEvent.java:64]: Since we are using log4j as backend, it seemed good to leverage logging levels. If confusing we can call `log` for  info/warn/error. However `debug` should be left for noisy diagnostic events.

File: core/client/fs/src/main/java/alluxio/client/file/AlluxioEvent.java (hunk 2 above)
Comment (ggezer) [reply core/client/fs/src/main/java/alluxio/client/file/AlluxioEvent.java]: camel casing is better.

File: core/common/src/main/java/alluxio/AlluxioEvent.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    mMsgFormat = msgFormat;
+  }
+
+  /**
+   * Fires the event.
+   *
+   * @param args arguments to event message string
+   */
+  public void fire(Object... args) {
+    String msgString = String.format(mMsgFormat, args);
```
Comment (gpang) [inline core/common/src/main/java/alluxio/AlluxioEvent.java:65]: I feel like instead of treating the event message as a format, we should just use the input as additional info/text for the event.

File: core/common/src/main/java/alluxio/AlluxioEvent.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        LOG.info(eventString);
+        break;
+      case ERROR:
+        LOG.error(eventString);
+        break;
+      case WARNING:
+        LOG.warn(eventString);
+        break;
+      case DIAGNOSTIC:
+        LOG.debug(eventString);
```
Comment (gpang) [inline core/common/src/main/java/alluxio/AlluxioEvent.java:78]: Isn't the event log going to a separate file?

In any case, events should always be logged, so I don't think any event should be at debug level. The goal for events is to see every event that occurs. Events logging should just not be noisy or extraneous.

File: core/common/src/main/java/alluxio/AlluxioEvent.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Defines and Manages well-known Alluxio events.
+ *
+ * !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+ * !!! - Keep increasing the event-id for new events. !!!
+ * !!! - Don't reassign event ids.                    !!!
```
Comment (gpang) [inline core/common/src/main/java/alluxio/AlluxioEvent.java:22]: We need a unit test which inspects all ids programmatically to validate there are no duplicates.

File: core/common/src/main/java/alluxio/AlluxioEvent.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+package alluxio;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Defines and Manages well-known Alluxio events.
+ *
+ * !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+ * !!! - Keep increasing the event-id for new events. !!!
```
Comment (gpang) [inline core/common/src/main/java/alluxio/AlluxioEvent.java:21]: I don't think this is as important, since it may make sense to partition the id space for different types of events. Like if we do event logging on the worker, maybe those are in a different segment of the id space.

File: core/common/src/main/java/alluxio/AlluxioEvent.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+ * !!! - Don't reassign event ids.                    !!!
+ * !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+ */
+public enum AlluxioEvent {
+  // Alluxio master events.
+  MasterProcessStarting(1000, EventType.INFO, "Master process starting."),
+  JournalSystemStarted(1001, EventType.INFO, "Journal system started. %s"),
+  MasterIsPrimary(1002, EventType.INFO, "Master process is now the primary."),
```
Comment (gpang) [inline core/common/src/main/java/alluxio/AlluxioEvent.java:29]: We may need to be more precise about these events. For example, I'd imagine the sequence of events could be the following. I don't know how these events map to this sequence:
process started -> journal started -> primacy gained -> serving started -> ... -> primacy lost -> serving stopped -> process stopped

Also, something like "MasterIsPrimary" seems more like a state instead of an event. This could be "primacy gained" if that is what this means.

We may also want to be precise of past tense and present tense. For example, "process starting" is different from "process started". I feel like past tense might be more clear what it means. Like, "journal catchup started" -> "journal catchup finished".

Comment (ggezer) [reply core/common/src/main/java/alluxio/AlluxioEvent.java]: Incorporated these to new event names. PTAL.

**Comment by ggezer** [comment]
alluxio-bot, test this please

File: core/common/src/main/java/alluxio/AlluxioEvent.java [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.stream.Collectors;
+
+/**
+ * Defines and Manages well-known Alluxio events.
+ */
+public enum AlluxioEvent {
+  // [1000,2000) Alluxio process events.
```
Comment (gpang) [inline core/common/src/main/java/alluxio/AlluxioEvent.java:24]: does this include job processes too?

File: core/common/src/main/java/alluxio/AlluxioEvent.java [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+  MasterProcessCreated(1000, "Alluxio master process created."),
+  JournalSystemStarted(1001, "Journal system started."),
```
Comment (gpang) [inline core/common/src/main/java/alluxio/AlluxioEvent.java:26]: Will there be an example of when the message is different from the name of the enum? It seems like every event so far has a message which just mirrors the enum name.

File: core/server/common/src/main/java/alluxio/master/AbstractPrimarySelector.java [hunk 13]
```diff
@@ -50,6 +51,8 @@
   private State mState = State.SECONDARY;
 
   protected final void setState(State state) {
+    AlluxioEvent.MasterIsTransitioning
+        .fire(String.format("Old-State: %s, New-State: %s", mState, state));
```
Comment (gpang) [inline core/server/common/src/main/java/alluxio/master/AbstractPrimarySelector.java:55]: fire already joins multiple arguments, so should this be split up into multiple strings?

File: core/common/src/main/java/alluxio/AlluxioEvent.java [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+
+  /**
+   * Fires the event.
+   *
+   * @param additionalInfo additional information to include in the event
+   */
+  public void fire(Object... additionalInfo) {
+    String eventString = String.format("Id: %d, Name: %s, Message:%s, AdditionalInfo: %s",
+        mId, this.name(), mMessage,
+        Arrays.stream(additionalInfo).map(Object::toString).collect(Collectors.joining(",")));
```
Comment (gpang) [inline core/common/src/main/java/alluxio/AlluxioEvent.java:91]: I think there should be a space after the comma?

File: core/server/master/src/main/java/alluxio/master/backup/BackupLeaderRole.java [hunk 15]
```diff
@@ -183,6 +185,7 @@ public BackupStatus backup(BackupPRequest request, StateLockOptions stateLockOpt
       // Store backup id to query later for async requests.
       backupId = mBackupTracker.getCurrentStatus().getBackupId();
     }
+    AlluxioEvent.BackupStarted.fire(backupId);
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/backup/BackupLeaderRole.java:188]: should this have a label for the id, like `String.format("backupId: %d", backupId)`

File: core/server/master/src/main/java/alluxio/master/backup/BackupTracker.java [hunk 16]
```diff
@@ -221,8 +222,10 @@ private void signalIfFinished() {
 
     if (mBackupStatus.isCompleted()) {
       mCompletion.set(null);
+      AlluxioEvent.BackupFinished.fire(mBackupStatus.getBackupId());
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/backup/BackupTracker.java:225]: we should label the int values in the output

File: core/common/src/main/java/alluxio/AlluxioEvent.java [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+    return mMessage;
+  }
+
+  /**
+   * Fires the event.
+   *
+   * @param additionalInfo additional information to include in the event
+   */
+  public void fire(Object... additionalInfo) {
+    String eventString = String.format("Id: %d, Name: %s, Message:%s, AdditionalInfo: %s",
```
Comment (gpang) [inline core/common/src/main/java/alluxio/AlluxioEvent.java:89]: How do you feel about an output like "EventName(EventId): Message. AdditionalInfo"? (We may or may not need message) Being concise could be more readable.

File: core/common/src/main/java/alluxio/AlluxioEvent.java (hunk 11 above)
Comment (ggezer) [reply core/common/src/main/java/alluxio/AlluxioEvent.java:24]: i didn't intend to.

File: core/common/src/main/java/alluxio/AlluxioEvent.java (hunk 12 above)
Comment (ggezer) [reply core/common/src/main/java/alluxio/AlluxioEvent.java:26]: yeah seems redundant till now.

**Comment by ggezer** [comment]
jenkins, test this please

**Comment by ggezer** [comment]
@gpang Added log4J configuration and reallocated event ids.

File: core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java [hunk 18]
```diff
@@ -666,4 +670,10 @@ public void format() throws IOException {
   public PrimarySelector getPrimarySelector() {
     return mPrimarySelector;
   }
+
+  @Override
+  public String toString() {
+    return String.format("RaftJournalSystem, in quorum: {%s}", mConf.getClusterAddresses().stream()
```
Comment (gpang) [inline core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java:676]: Should we just use `MoreObjects.toStringHelper()` for this?

File: core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournalSystem.java [hunk 19]
```diff
@@ -212,4 +216,9 @@ public void checkpoint() throws IOException {
       journal.checkpoint();
     }
   }
+
+  @Override
+  public String toString() {
+    return String.format("UfsJournalSystem, based at %s", mBase.toString());
```
Comment (gpang) [inline core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournalSystem.java:222]: Should we just use `MoreObjects.toStringHelper()` for this?

File: core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java [hunk 20]
```diff
@@ -805,9 +807,20 @@ private void freeSpaceInternal(long sessionId, long minContiguousBytes, long min
...
       LOG.error(
-          "Failed to free space. Min contiguous requested: {}, Min available requested: {}, "
+          "Failed to free space in session: {}. "
+              + "Min contiguous requested: {}, Min available requested: {}, "
               + "Blocks iterated: {}, Blocks removed: {}, " + "Space freed: {}",
-          minContiguousBytes, minAvailableBytes, blocksIterated, blocksRemoved, spaceFreed);
+          sessionId, minContiguousBytes, minAvailableBytes,
+          blocksIterated, blocksRemoved, spaceFreed);
+
+      AlluxioEvent.BlockStoreEvictionFailed.fire(
```
Comment (gpang) [inline core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java:816]: Is it possible that this is a noisy event? Would this be considered an event?

File: conf/log4j.properties [hunk 21]
```diff
@@ -181,5 +181,16 @@ log4j.appender.FUSE_LOGGER.MaxBackupIndex=10
 log4j.appender.FUSE_LOGGER.layout=org.apache.log4j.PatternLayout
 log4j.appender.FUSE_LOGGER.layout.ConversionPattern=%d{ISO8601} %-5p %c{1} - %m%n
 
+# Appender for Alluxio events
+log4j.appender.EVENT_LOGGER=org.apache.log4j.RollingFileAppender
+log4j.appender.EVENT_LOGGER.File=${alluxio.logs.dir}/events.log
```
[the remark this answers is not in the document]
Comment (ggezer) [reply conf/log4j.properties:186]: This is mostly to aid production cluster investigations.
Even when not, I think seeing logs in a single file is not a problem.

**Comment by ggezer** [comment]
jenkins, test this please

File: conf/log4j.properties (hunk 21 above)
[the remark this answers is not in the document]
Comment (gpang) [reply conf/log4j.properties:186]: Is it possible/correct for multiple processes to write to the same log file?

[the remark this answers is not in the document]
Comment (ggezer) [reply conf/log4j.properties:186]: I removed the worker events as I already have the exact same information in worker logs.
We can later figure out the plumbing for log4j If we want to emit events from other processes.

File: core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java [hunk 22]
```diff
@@ -666,4 +671,12 @@ public void format() throws IOException {
   public PrimarySelector getPrimarySelector() {
     return mPrimarySelector;
   }
+
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this)
+        .add("Quorum-Members", mConf.getClusterAddresses().stream()
```
Comment (gpang) [inline core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java:678]: many other labels don't have dashes. Should this be "QuorumMembers"?

File: core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournalSystem.java [hunk 23]
```diff
@@ -212,4 +217,11 @@ public void checkpoint() throws IOException {
       journal.checkpoint();
     }
   }
+
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this)
+        .add("Ufs-Uri", mBase)
```
Comment (gpang) [inline core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournalSystem.java:224]: should this be "UfsUri"?

File: core/server/master/src/main/java/alluxio/master/file/activesync/ActiveSyncer.java [hunk 24]
```diff
@@ -78,6 +80,7 @@ public ActiveSyncer(FileSystemMaster fileSystemMaster, ActiveSyncManager syncMan
   @Override
   public void heartbeat() {
     LOG.debug("start sync heartbeat for {} with mount id {}", mMountUri, mMountId);
+    AlluxioEvent.ActiveSyncStarted.fire();
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/activesync/ActiveSyncer.java:83]: I feel like these active sync events will be noisy. Do we need them now?

File: core/server/common/src/main/java/alluxio/master/AbstractPrimarySelector.java [hunk 25]
```diff
@@ -50,6 +52,8 @@
   private State mState = State.SECONDARY;
 
   protected final void setState(State state) {
+    AlluxioEvent.MasterIsTransitioning.fire(
+        new Pair<>("Old-State", mState), new Pair<>("New-State", state));
```
Comment (gpang) [inline core/server/common/src/main/java/alluxio/master/AbstractPrimarySelector.java:56]: should this be "OldState" and "NewState"?

File: core/server/master/src/main/java/alluxio/master/file/activesync/ActiveSyncer.java (hunk 24 above)
Comment (ggezer) [reply core/server/master/src/main/java/alluxio/master/file/activesync/ActiveSyncer.java:83]: why theyb would be noisy?

**Comment by ggezer** [comment]
jenkins, test this please

File: core/server/master/src/main/java/alluxio/master/file/activesync/ActiveSyncer.java (hunk 24 above)
Comment (gpang) [reply core/server/master/src/main/java/alluxio/master/file/activesync/ActiveSyncer.java:83]: Isn't it syncing on a heartbeat? We would see this event on every tick, which could be frequent and noisy. I don't think we want an event associated with a heartbeat tick.

File: core/common/src/main/java/alluxio/AlluxioEvent.java [hunk 26]
```diff
@@ -0,0 +1,92 @@
...
+ */
+public enum AlluxioEvent {
+  // Alluxio events.
+  // Next id:1019.
+  MasterProcessCreated(1000),
+  JournalSystemStarted(1001),
+  JournalSystemStopped(1002),
+  JournalSystemGainedPrimacy(1003),
+  JournalSystemLostPrimacy(1004),
+  MasterRpcServerStarted(1005),
```
Comment (gpang) [inline core/common/src/main/java/alluxio/AlluxioEvent.java:35]: Are these related to `MasterProcessCreated` event? Should they be grouped together?

Comment (ggezer) [reply core/common/src/main/java/alluxio/AlluxioEvent.java:35]: grouped how?

File: core/server/master/src/main/java/alluxio/master/file/activesync/ActiveSyncer.java (hunk 24 above)
Comment (ggezer) [reply core/server/master/src/main/java/alluxio/master/file/activesync/ActiveSyncer.java]: removed events.
